src/platform_impl/linux/kms/event_loop.rs

```diff
@@ -0,0 +1,1461 @@
+                file!(),
+                platform_impl::OsError::KmsMisc("failed to find suitable GPU"),
+            )
+        })
+}
+
+pub struct EventLoop<T: 'static> {
+    /// Event loop.
+    event_loop: calloop::EventLoop<'static, EventSink>,
+
+    /// Pending user events.
+    pending_user_events: Rc<RefCell<Vec<T>>>,
+
+    /// Sender of user events.
+    user_events_sender: calloop::channel::Sender<T>,
+
+    /// Window target.
+    window_target: event_loop::EventLoopWindowTarget<T>,
+}
+
+impl<T: 'static> EventLoop<T> {
+    pub fn new() -> Result<EventLoop<T>, OsError> {
+        #[cfg(feature = "kms-ext")]
+        // When we create the seat here, we should probably wait for it to become active before we
+        // use it.
+        let mut seat = {
+            use std::sync::atomic::Ordering;
+            // Allows us to know when the seat becomes active
+            let active = Arc::new(AtomicBool::new(false));
+            let t_active = active.clone();
+            let mut s = libseat::Seat::open(
+                move |_, event| {
+                    if let libseat::SeatEvent::Enable = event {
+                        t_active.store(true, Ordering::SeqCst);
+                    }
+                },
+                None,
+            )
+            .map_err(|e| {
+                OsError::new(
+                    line!(),
+                    file!(),
+                    platform_impl::OsError::KmsError(format!("failed to open libseat: {}", e)),
+                )
+            })?;
+
+            // While our seat is not active dispatch it so that the seat will activate
+            while !active.load(Ordering::SeqCst) {
+                s.dispatch(-1).map_err(|e| {
+                    OsError::new(
+                        line!(),
+                        file!(),
+                        platform_impl::OsError::KmsError(format!("failed to dispatch seat: {}", e)),
+                    )
+                })?;
+            }
+            s
+        };
+
+        #[cfg(feature = "kms-ext")]
+        // Safety
+        //
+        // This string value has the same lifetime as the seat in question, and will not be dropped
+        // until the seat is, which is not before `udev_assign_seat` is run.
+        let seat_name = unsafe { std::mem::transmute::<&str, &'static str>(seat.name()) };
+        #[cfg(not(feature = "kms-ext"))]
+        let seat_name = "seat0";
+
+        // find_card_path uses `udev` to enumerate the cards that are currently available, and then
+        // choose the first (usually perferred) one
+        let card_path = std::env::var("WINIT_DRM_CARD")
+            .ok()
+            .map_or_else(|| find_card_path(seat_name), |p| Ok(Into::into(p)))?;
+
+        #[cfg(feature = "kms-ext")]
+        // Opening the card using our seat allows us to do so unprivallaged
+        let dev = seat
+            .open_device(&card_path)
+            .map_err(|e| {
+                OsError::new(
```

Comment:
There's a macro `os_error!` for that already in `src/error.rs`

Reply:
Fixed